### ISSUE TITLE
Placement of new window when screen is full

### ISSUE THREAD

**Issue by andrei-pavel** [lingtjien/Grid-Tiling-Kwin#48]
Hello,

Thank you very much for this awesome script!

Does it control where new windows are placed?

With multiple screens, if the focused screen is full of windows e.g. 6/6 and if there is still space on a second screen, a new window is created in a new workspace instead of creating it on a different screen in the same workspace. I find the latter to be preferable. Is this intended? Is this in the script's scope?

**Comment by lingtjien** [comment]
I'm not completely following along with you...

# Intended behavior

regarding multi-screen setups:

- fill the active screen and virtual desktop with clients until it is full
- when full -> attempt the next physical screen until it's filled with clients
- when the next physical screen is full -> attempt the next virtual desktop on the first screen
- when all virtual desktops and physical screens are full, float all subsequent clients

# Example 

assuming 2 virtual desktops and 2 physical screens with the active virtual desktop being 1 and physical screen 1

fill order: virtual desktop 1, screen 1 -> virtual desktop 1, screen 2 -> virtual desktop 2, screen 1 -> virtual desktop 2, screen 2

To be fair, I haven't tested the multiscreen code in a long time, since my setup only has my laptop screen, so if this is not the behavior you observe then I'll have to take a look.

**Comment by andrei-pavel** [comment]
In my case, it behaves according to the fill order you specified. But I ask: what if virtual desktop 1, screen 2 is full and all the others are not (or empty for simplicity), including virtual desktop 1, screen 1 which plays an essential role in this situation. Should the new window appear in virtual desktop 1, screen 1? In my situation it appears in virtual desktop 2, screen 1.

The general question is, should you first fill previous screens of the same virtual desktop or keep it simple and always advance to the next virtual desktop?

I understand it could be done both ways now, but this is how I thought it was not the most intuitive approach: I keep screen 2 on top of screen 1, so numbering from top to bottom I thought they were reversed in terms of ordinality and was tricked thinking that from virtual desktop 1, screen 1 it goes all the way to virtual desktop 2, screen 2, but that's not the case.

You can close this issue, but I do think it would be better if it was changed and it doesn't have to be slow, maybe you can keep track of what screen needs to be filled next and not always search it from the beginning. What do you think?

**Comment by lingtjien** [comment]
Ah okay, so you want the algorithm to prioritize filling screens and then go to the next virtual desktop once they're all full.

The current way was the most straightforward to implement, so I went with that but I agree that prioritizing screens is more elegant. I'll see what I can do :)

**Comment by lingtjien** [comment]
I tried to implement it in the dev branch. The goal I had in mind was the following priority:

1. clients tile to desktop and screen that they are started on
2. next available screen
3. next available desktop

**Comment by andrei-pavel** [comment]
Sorry, mail got lost in my inbox.

Also sorry, but current develop is broken for me.

- Refresh shortcut doesn't work.
- New windows don't automatically tile anymore.

These are consistent and deterministic. If I go back to master, all of the above work as before.

The way I switch is:
```
git checkout "${branch}"
kpackagetool5 --type KWin/Script -i "${directory}"
kwin_x11 --replace
```

Tried it multiple times, also with reboot instead of `kwin_x11 --replace`, simptoms are the same so I think I'm testing properly.

**Comment by lingtjien** [comment]
There is a major change I made to the grid size specification. You **must** now supply a grid for every screen (and additional screens that you plan to connect).

So what are the values you have for `row` and `column` and how many physical screens do you have connected?

**Comment by andrei-pavel** [comment]
Two phyisical screens, two rows, three columns

![image](https://user-images.githubusercontent.com/6838492/73613494-eee1b300-45fe-11ea-9688-7d3074e5b3a8.png)

**Comment by andrei-pavel** [comment]
Right now, I fill one screen with 6 windows. When I spawn the 7th window, it spawns non-tiled on a random position on that same screen. Before the 7th screen, tiling worked & Grid-TIling shortcuts like Refresh worked. After the 7th screen, tiling & Grid-Tiling shortcuts don't work anymore. It's maybe like it crashes? Can I enter debug mode and run code line by line in some way?

**Comment by lingtjien** [comment]
Yeah so the script will crash if you do not define a row and column size for your physical screen.

So right now you have the values `2` and `3` for `row` and `column` so right now the script will crash if you have more than 1 screen because you only supplied a grid size for 1 screen. (Previously it would take the grid size of the first screen for all your additional screens, but I removed that because it requires a lot of extra code only to handle that)

So if your system has 2 screens normally and you plan to add an extra screen occasionally, then you **must** give the script 3 grid sizes, so 3 values for both the `row` and `column` variables (separated by a comma).

TLDR: change the values for `row` and `column` under the section `Grid Sizes` to `2, 2` and `3, 3` or if you plan to add a third screen occasionally `2, 2, 2` and `3, 3, 3`. Assuming that you want the same grid size for every screen. If you want a different grid size on a certain screen you can do `2, 2, 3` and `3, 3, 4` then the third screen will have a grid of 3x4 while the first and second screen both have a grid of 2x3 and 2x3

**Comment by andrei-pavel** [comment]
Thank you! It makes sense. Settings button seems to have disappeared from the kwin scripts menu on my Ubuntu at work. Will try on Arch from home.

**Comment by lingtjien** [comment]
the settings button is only visible when you have the correct file in `/kservies5`. See the last 3 commands in the installation procedure (both normal and local are the same).

**Comment by andrei-pavel** [comment]
I'm guilty of being ignorant of that part of the documentation. I had consciously removed it from my script saying "ah, this does nothing".

Tested develop and it's working accordingly. Thank you very much!
